### PR TITLE
Coroutine stackoverflow detection for linux/posix; Issue #2408

### PR DESCRIPTION
**Pull request by ct-clmsn**
Initial implementation of stackoverflow detection for coroutines issue tracker [#2408](https://github.com/STEllAR-GROUP/hpx/issues/2408) . The source was derived from posix/linux system man pages, the [rethinkdb link](https://rethinkdb.com/blog/handling-stack-overflow-on-custom-stacks/) provided in the issue conversation thread, and a [blog post](http://evanjones.ca/software/threading.html) about coroutine implementations.

### CODE REVIEW

**Comment by ct-clmsn** [comment]
@hkaiser mixed up my unit test with an existing unit test.

File: hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp [hunk 1]
```diff
@@ -140,7 +151,19 @@ namespace hpx { namespace threads { namespace coroutines
 
             x86_linux_context_impl()
                 : m_stack(nullptr)
-            {}
+            {
+                segv_stack.ss_sp = valloc(SEGV_STACK_SIZE);
```
Comment (sithhell) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: Could we have that depending on a define such that is configurable through cmake?

Comment (ct-clmsn) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: @sithhell sure thing. How would you want to distinguish or designate the cmake variables? example: HPX_COROUTINE_STACKOVERFLOW_DETECTION

**Comment by ct-clmsn** [comment]
@ABresting I'm aware of the GPL concerns mentioned in issue #2408. I did not read and have not read the libsegsegv implementation source files. At the top of this PR's commentary thread, I noted the reference materials used to implement this PR. I'd strongly encourage reviewing the links provided at the top of this PR commentary thread if you have concerns regarding the origin of this PR's implementation.

File: hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp (hunk 1 above)
Comment (ct-clmsn) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: @sithhell is streaming to std::cerr acceptable or would ya'll prefer hpx::cerr?

Comment (hkaiser) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: We usually name our cmake config variable `HPX_WITH_<FOO>` and the corresponding preprocessor constant `HPX_HAVE_<FOO>`.

Also, `hpx::cerr` might not really work as it would send things to the console which might not be operational under stack-overflow conditions.

Comment (ct-clmsn) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: @sithhell just placed in the cmake modification you suggested - hope that's sufficient. thanks! @hkaiser thank you for the guidance!

File: hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp [hunk 2]
```diff
@@ -140,7 +155,23 @@ namespace hpx { namespace threads { namespace coroutines
 
             x86_linux_context_impl()
                 : m_stack(nullptr)
+#if defined(HPX_HAVE_THREAD_STACKOVERFLOW_DETECTION)
+            {
+                segv_stack.ss_sp = valloc(SEGV_STACK_SIZE);
+                segv_stack.ss_flags = 0;
+                segv_stack.ss_size = SEGV_STACK_SIZE;
+
+                bzero(&action, sizeof(action));
```
Comment (hkaiser) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: May I suggest to use `std::memset(&action), '\0', sizeof());` instead of `bzero` (also `#include <cstring>` for memset to be declared)? `bzero` is not portable (not that we would need for it to be portable) and not part of the std C++ library.

Comment (ct-clmsn) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: @hkaiser sure thing! modifications made and pushed.

**Comment by ABresting** [comment]
@ct-clmsn rethink-db's method is under Apache licensing, so even I am not so sure how safe it will be.

**Comment by biddisco** [comment]
The apache license allows redistribution under a new license, providing the original code keeps the apache license, so it would be compatible with HPX if any apache licensed code has been used directly.

from https://www.whitesourcesoftware.com/whitesource-blog/top-10-apache-license-questions-answered/

"You can choose to release the modified or derived products under different licenses, the unmodified parts of the software, however, must retain the Apache License, and you cannot name your modified version in any way that suggests that the final product is endorsed or created by the ASF.

Additionally, if you want to add a copyright statement about all the modifications that you’ve done to any Apache licensed software; you are free to do so. Since the Apache License doesn’t require you to release the modified code under the same license, you can choose to add specific license terms and conditions that govern how others use, reproduce, or distribute your modified code."

**Comment by ct-clmsn** [comment]
@ABresting the implementation of this PR used tutorial information (examples) provided in the two blog posts (not licensed program source code) mentioned in my top comment, combined with my interpretation of posix/linux manual pages. @biddisco thanks!

**Comment by biddisco** [comment]
So to finish my thought on the subject - if we were to use the code, add the boost licence, add a note to say that the code was inspired by the blog posts #link1, #link2, &etc then since the author has not looked at the other library, it qualifies as not being a derivative work of the GPL'd code. It could be construed as a derivative of some apache stuff and we therefore note that, include the apache license and refer again to the articles, then I think we're ok. If anyone ever raises an objection, the code can be pulled until the issue is resolved - at that time.

**Comment by hkaiser** [comment]
@biddisco: sounds like a plan. Better beg for forgiveness afterwards than to ask for permission upfront ;-)

**Comment by ct-clmsn** [comment]
@biddisco @hkaiser web-link-resources have been added to the source in this PR in a comment; *pretty sure* all of the extra EOL character spaces have been removed.

**Comment by hkaiser** [comment]
Thanks Chris. Code-wise this looks good now. I will try it out asap (I think @sithhell wanted to try it out as well).

**Comment by ct-clmsn** [comment]
Great! Thank you! Should the error message include a suggestion about problem decomposition?

File: CMakeLists.txt [hunk 3]
```diff
@@ -139,6 +139,14 @@ if("${CMAKE_SYSTEM_NAME}" STREQUAL "Linux")
 #    set(DEFAULT_MALLOC "jemalloc")
 endif()
 
+hpx_option(HPX_WITH_THREAD_STACKOVERFLOW_DETECTION
+  BOOL
+  "Enable stackoverflow detection for HPX threads/coroutines. (default: ON)"
+  ON ADVANCED)
```
Comment (sithhell) [inline CMakeLists.txt]: Should we make this OFF by default?

Comment (hkaiser) [inline CMakeLists.txt]: I'd suggest leaving it `ON` by default at least in Debug mode.

Comment (hkaiser) [inline CMakeLists.txt]: Also, would it make sense to make enable/disable a runtime property?

File: tests/unit/threads/CMakeLists.txt [hunk 4]
```diff
@@ -14,6 +14,7 @@ set(tests
     thread_launching
     thread_mf
     thread_stacksize
+    thread_stacksize_overflow
     thread_suspension_executor
     thread_yield
    )
```
Comment (sithhell) [inline tests/unit/threads/CMakeLists.txt]: The CMakeLists.txt here need to flag the added test as "will fail". I suggest something like this:
```
set_tests_properties(tests.unit.threads.thread_stacksize_overflow PROPERTIES
  PASS_REGULAR_EXPRESSION "Stack overflow in coroutine at address 0x[0-9a-fA-F]*")
```

This will mark the test as passed when it fails with that output. That would require to add this test only when the overflow detection is activated.

Comment (hkaiser) [inline tests/unit/threads/CMakeLists.txt]: Also, this should be enabled on Linux-based systems only.

File: hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp [hunk 5]
```diff
@@ -189,8 +228,49 @@ namespace hpx { namespace threads { namespace coroutines
...
+                std::cerr << std::endl << "Configure the hpx runtime to "
+                    << "allocate a larger "
+                    << "coroutine stack size." << std::endl
+                    << "Use the hpx.stacks.small_size, "
+                    << "hpx.stacks.medium_size, "
+                    << std::endl
+                    << "hpx.stacks.large_size, "
+                    << "or hpx.stacks.huge_size runtime "
+                    << std::endl
+                    << "flags to configure coroutine heap sizes."
```
Comment (sithhell) [inline hpx/runtime/threads/coroutines/detail/context_linux_x86.hpp]: This should read "stack sizes"

File: hpx/runtime/threads/coroutines/detail/context_posix.hpp [hunk 6]
```diff
@@ -223,8 +239,55 @@ namespace hpx { namespace threads { namespace coroutines
...
+                std::cerr << std::endl << "Configure the hpx runtime to "
+                    << "allocate a larger "
+                    << "coroutine stack size." << std::endl
+                    << "Use the hpx.stacks.small_size, "
+                    << "hpx.stacks.medium_size, "
+                    << std::endl
+                    << "hpx.stacks.large_size, "
+                    << "or hpx.stacks.huge_size runtime "
+                    << std::endl
+                    << "flags to configure coroutine heap sizes."
```
Comment (sithhell) [inline hpx/runtime/threads/coroutines/detail/context_posix.hpp]: same as above

File: CMakeLists.txt (hunk 3 above)
Comment (sithhell) [inline CMakeLists.txt]: I am not sure about the overheads this detection adds. I am leaning towards only enabling it on request (If posix/linux based platform), given that prerequiste, we can also safely enable it when the build type is debug.

Comment (ct-clmsn) [inline CMakeLists.txt]: @sithhell @hkaiser just to clarify: disable by default and enable for debug builds by default?

File: tests/unit/threads/CMakeLists.txt (hunk 4 above)
Comment (ct-clmsn) [inline tests/unit/threads/CMakeLists.txt]: @sithhell thank you for the suggestion. @hkaiser will be sure to provide a platform check.

Comment (hkaiser) [inline tests/unit/threads/CMakeLists.txt]: @ct-clmsn I will do that. Let's go ahead with this as is.

File: CMakeLists.txt (hunk 3 above)
Comment (sithhell) [inline CMakeLists.txt]: @ct-clmsn Enable for debug builds by default, Disable by default for the remaining build types, I'd say.

Comment (hkaiser) [inline CMakeLists.txt]: @sithhell: we were not able to resolve a similar configuration issue a while back (see: #2365). How would you suggest to do this (if you had a new idea we could go back and fix #2365 as well...)?

Comment (sithhell) [inline CMakeLists.txt]: The issue we had back then was to supporting multi configuration builds. As those don't exist in any of the *nix ecosystems (or does WSL change the picture?), I would consider this a non problem for this particular feature.

Comment (hkaiser) [inline CMakeLists.txt]: Ahh right, I withdraw my comment, then...

**Comment by ct-clmsn** [comment]
@hkaiser thank you for helping to finalize the commit!
